### ISSUE TITLE
O=note shoots a note twice, and can't output two notes

### ISSUE THREAD

**Issue by aaronventure** [jtackaberry/reaticulate#20]
I seem to be getting an issue where I want to use Berlin Woodwinds and with Inspector I had a keyswitch for the patch with no vibrato in Capsule, and I would then route it with a JS note filter and mapper to also trigger another keyswitch at the same time - allowing me to crossfade between the no vib and molto vib patches. 

![image](https://user-images.githubusercontent.com/34456677/33853627-dac02e46-debe-11e7-8273-62a5483ea571.png)

When I try the same with Reaticulate, it seems to shoot two notes of the keyswitch, because in Capsule I see  4 patches now loaded for crossfading, all the same ones
![image](https://user-images.githubusercontent.com/34456677/33853568-b3f09d28-debe-11e7-9b9c-b7b0c2ad6356.png)
which is something I can't replicate by manually keyswitching on my keyboard - I guess it has to be fast. Also, after that, once I swtich to another articulation, it puts it at the end of the 4-selection chain.

![image](https://user-images.githubusercontent.com/34456677/33853547-a1467bc0-debe-11e7-8fc5-eb1cb9ea66b1.png)

And the first three notes seem to be hanging - playing and pausing resets the whole thing and selecting another articulation after the reset works fine - UNLESS it's the Legato articulation (which should trigger another one.)

I realized that Reaticulate could have a built in solution for this by allowing me to trigger two notes, but o=note24/note:25 doesn't seem to be working - it only triggers the last one. Are they shot at the same time/playing back at the same time?

There's a workaround for my situation, but it'd be great if I could have it this way.

**Comment by aaronventure** [comment]
UPDATE: This probably confirms that it shoots two notes: Using JS MIDI Duplicate Note Filter solves the double double 
![image](https://user-images.githubusercontent.com/34456677/33854936-afda59ae-dec3-11e7-8375-4968cdc8332f.png)

But the problem with these two notes hanging still persists.

**Comment by jtackaberry** [comment]
Can you insert a ReaControlMIDI in between Reaticulate and Kontakt and use the log panel to verify exactly what Reaticulate is sending to Kontakt, for both the single note case, and the double note case (e.g. o=note24/note:25)?  This will help quite a lot with troubleshooting. :)

**Comment by jtackaberry** [comment]
Oh, and also provide the full bank file you're using?

**Comment by aaronventure** [comment]
Wow, I never realized ReaControlMIDI had that option. Fantastic. God, I love Reaper.

For every articulation it's this
![image](https://user-images.githubusercontent.com/34456677/33857721-2fe1ca52-decd-11e7-998b-439bc1adae5f.png)

And after the Note reroute it's this
![image](https://user-images.githubusercontent.com/34456677/33857811-9997bc5e-decd-11e7-85fe-61fe50e3c14c.png)

After the Duplicate note filter it's this for the LEGATO which goes to the reroute
![image](https://user-images.githubusercontent.com/34456677/33857848-c49b132e-decd-11e7-8763-c0950c97763b.png)

But it's the same for all other articulations - still double notes. However, the two legato notes that come out of the rerouter seem to be hanging as clicking the articulation again does nothing, but it'll send once after I play and pause.

Here's the bank file
[bww.txt](https://github.com/jtackaberry/reaticulate/files/1549646/bww.txt)

**Comment by aaronventure** [comment]
And here's with
![image](https://user-images.githubusercontent.com/34456677/33859612-fe400c5e-ded4-11e7-8ee6-c09399c27100.png)

It sends
![image](https://user-images.githubusercontent.com/34456677/33859601-f4591ce4-ded4-11e7-8a46-76696f9d23c6.png)
But this time BWW only receives the last one
![image](https://user-images.githubusercontent.com/34456677/33859626-1b2c4468-ded5-11e7-8d89-bb280b3241fa.png)

**Comment by blawler** [comment]
I would like to see this feature implemented as well.  I am having a similar issue in Palette for Kontakt where I want to use combined articulations (staccato+tenuto) The note-hold command still invokes a note-off MIDI event just before the next output action.  I need it to simulate holding two keys down at the same time. If the last "note-hold" sent it's note off event AFTER THE NEXT "note" command sent a note on event, it would work for me.  Would that screw it up for other applications?

Great utility BTW- thank you.  I have my articulation library "finished" for Red Room Audio / Palette with BrushPack 1, except for these "custom articulations" that I would like to add.

**Comment by jtackaberry** [comment]
Ah, yes, thank you for helping me make sense of "past me."  I believe this is indeed exactly the thing I had intended.  Future me misinterpreted the "multiple held keyswitches" comment to refer to a held note per group, but in fact the "past me" comment was added before I introduced groups into the code.

Fixing this will require a lot more work than just removing send_deferred_note_off() which, as you've done it, will undoubtedly end up with stuck notes (i.e. a note-on that never receives a note-off).  Minimally I'll have to expand held_keyswitches to track multiple notes per channel per group, and then make emit_output_event() a bit cleverer.

@aaronventure  can you confirm this is what you had intended when you originally opened the issue, i.e. that given two note output events, both note-on events should be sent before either of the note-off events, simulating both notes being pressed at the same time?

**Comment by aaronventure** [comment]
@jtackaberry Yes. I'm sorry if I wasn't clear on that. 

It still triggers both articulations (although twice) for Berlin Woodwinds the way it is right now, but these two notes would then seem to hang. 

Were you able to confirm or recreate the double note issue?

**Comment by blawler** [comment]
@jtackaberry Yep, you understand completely.  By dropping the note off from the note-hold routine I have to hit the individual articulations (note events) to close the "hangers".  Might one approach would be to expand the held_keyswitches array with a dimension to encompass x number of notes?  Your routines/functions would need a loop to deal with them.  Again- I'm out of touch with programming paradigms these days.  The last book on the topic I read was Kernighan and Ritchie ;).  Thanks for looking into it!

**Comment by jtackaberry** [comment]
Related to #26 (which talks about note-hold) which I consider to be more a bug whereas this more a feature.  Still, I intend to solve them both at the same time as they involve the same area of the code.
